chat-sdk-core/src/main/java/co/chatsdk/core/base/BaseLocationAdapter.java

```diff
@@ -0,0 +1,104 @@
+package co.chatsdk.core.base;
+
+import android.Manifest;
+import android.app.Activity;
+import android.content.Context;
+import android.content.pm.PackageManager;
+import android.location.Location;
+import android.os.Looper;
+
+import com.google.android.gms.location.FusedLocationProviderClient;
+import com.google.android.gms.location.LocationCallback;
+import com.google.android.gms.location.LocationRequest;
+import com.google.android.gms.location.LocationResult;
+import com.google.android.gms.location.LocationServices;
+
+import java.util.List;
+
+import androidx.core.content.ContextCompat;
+import co.chatsdk.core.session.ChatSDK;
+import co.chatsdk.core.utils.DisposableList;
+import co.chatsdk.core.utils.PermissionRequestHandler;
+import io.reactivex.Completable;
+import io.reactivex.Observable;
+import io.reactivex.Single;
+
+/**
+ * Created by Pepe on 01/25/19.
+ */
+
+public class BaseLocationAdapter {
+
+    protected final FusedLocationProviderClient locationClient;
+    protected final DisposableList disposableList = new DisposableList();
+
+    protected Context context() {
+        return ChatSDK.shared().context();
+    }
+
+    public BaseLocationAdapter() {
+        locationClient = LocationServices.getFusedLocationProviderClient(context());
+    }
+
+    public Observable<Location> requestLocationUpdates(long interval, int distance) {
+        return requestLocationUpdates(interval)
+                .distinctUntilChanged((l1, l2) -> l1.distanceTo(l2) < distance);
+    }
+
+    public Observable<Location> requestLocationUpdates(long interval) {
+        LocationRequest locationRequest = new LocationRequest();
+        locationRequest.setInterval(interval * 1000);
+        locationRequest.setFastestInterval(interval * 1000);
+        locationRequest.setPriority(LocationRequest.PRIORITY_BALANCED_POWER_ACCURACY);
+        if (ContextCompat.checkSelfPermission(context(), Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ContextCompat.checkSelfPermission(context(), Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
+            return Observable.error(new Error("location permissions not granted"));
+        }
+        return Observable.create(observable -> {
+            LocationCallback locationCallback = new LocationCallback() {
+                @Override
+                public void onLocationResult(LocationResult locationResult) {
+                    Location location = getMostAccurateLocation(locationResult.getLocations());
+                    if (location != null) {
+                        observable.onNext(location);
+                    }
+                }
+            };
+            locationClient.requestLocationUpdates(locationRequest, locationCallback, Looper.myLooper());
+        });
```

Comment:
Need to add `.subscribeOn(Schedulers.single()).observeOn(AndroidSchedulers.mainThread());`

Reply:
I don't understand where I should add this, `requestLocationUpdates` doesn't return an observable.

Reply:
I think it does:

```
public Observable<Location> requestLocationUpdates(long interval) {
        return Observable.create(observable -> {
```

You can add it here:

```
            };
            locationClient.requestLocationUpdates(locationRequest, locationCallback, Looper.myLooper());
        }).subscribeOn(Schedulers.single()).observeOn(AndroidSchedulers.mainThread());
```